src/Akka.Streams.Kafka/Stages/CommitableConsumerStage.cs

```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Akka.Streams.Kafka.Messages;
+using Akka.Streams.Kafka.Settings;
+using Akka.Streams.Stage;
+using Confluent.Kafka;
+
+namespace Akka.Streams.Kafka.Stages
+{
+    internal class CommitableConsumerStage<K, V, Msg> : GraphStageWithMaterializedValue<SourceShape<Msg>, Task>
+    {
+        private readonly ConsumerSettings<K, V> _settings;
+        private readonly ISubscription _subscription;
+
+        protected readonly Outlet<Msg> Out = new Outlet<Msg>("out");
+        public override SourceShape<Msg> Shape { get; }
+
+        public CommitableConsumerStage(ConsumerSettings<K, V> settings, ISubscription subscription)
+        {
+            _settings = settings;
+            _subscription = subscription;
+            Shape = new SourceShape<Msg>(Out);
+        }
+
+        public override ILogicAndMaterializedValue<Task> CreateLogicAndMaterializedValue(Attributes inheritedAttributes)
+        {
+            return new LogicAndMaterializedValue<Task>(new KafkaCommitableSourceStage<K, V>(_settings, _subscription, Shape), Task.CompletedTask);
+        }
+    }
+
+    internal class KafkaCommitableSourceStage<K, V> : TimerGraphStageLogic
+    {
+        private readonly ConsumerSettings<K, V> _settings;
+        private readonly ISubscription _subscription;
+        private readonly Outlet _out;
+        private Consumer<K, V> _consumer;
+        private Action<Message<K, V>> _messagesReceived;
+        private Action<IEnumerable<TopicPartition>> _partitionsAssigned;
+        private Action<IEnumerable<TopicPartition>> _partitionsRevoked;
+
+        private const string TimerKey = "PollTimer";
+
+        private readonly Queue<CommittableMessage<K, V>> _buffer = new Queue<CommittableMessage<K, V>>();
+
+        public KafkaCommitableSourceStage(ConsumerSettings<K, V> settings, ISubscription subscription, Shape shape) : base(shape)
+        {
+            _settings = settings;
+            _subscription = subscription;
+            _out = shape.Outlets.FirstOrDefault();
+
+            SetHandler(_out, onPull: () =>
+            {
+                if (_buffer.Count > 0)
+                {
+                    Push(_out, _buffer.Dequeue());
+                }
+                else
```

Comment:
You should make this another if and check if buffer is empty, i.e. If the last element was send downstream immediately poll and don't wait for new demand .

Reply:
@Silv3rcircl3 is it possible to create an integration test for this case?

Reply:
You could put two messages into Kafka, request 1, wait for the timer Intervall, i.e. Buffer = 1 and therefore timer isn't scheduled again, put another message into Kafka, request another one ( buffer is not empty therefore onPull will currently only push without polling Kafka ), delete message from Kafka, request another one, buffer is empty -> tries to pull but Kafka is empty too so test will fail. Test should pass if you check onPull if buffer is empty after pushing an element.